gax-grpc/src/main/java/com/google/api/gax/grpc/GrpcCallContext.java

```diff
@@ -170,12 +183,34 @@ public GrpcCallContext withStreamIdleTimeout(@Nullable Duration streamIdleTimeou
     }
 
     return new GrpcCallContext(
-        channel, callOptions, streamWaitTimeout, streamIdleTimeout, channelAffinity);
+        channel, callOptions, streamWaitTimeout, streamIdleTimeout, channelAffinity, extraHeaders);
   }
 
+  @BetaApi("The surface for channel affinity is not stable yet and may change in the future.")
   public GrpcCallContext withChannelAffinity(@Nullable Integer affinity) {
     return new GrpcCallContext(
-        channel, callOptions, streamWaitTimeout, streamIdleTimeout, affinity);
+        channel, callOptions, streamWaitTimeout, streamIdleTimeout, affinity, extraHeaders);
+  }
+
+  @BetaApi("The surface for extra headers is not stable yet and may change in the future.")
+  @Override
+  public GrpcCallContext withExtraHeaders(@Nullable Map<String, List<String>> extraHeaders) {
+    ImmutableListMultimap.Builder<String, String> newExtraHeadersBuilder =
+        ImmutableListMultimap.builder();
+    if (extraHeaders != null) {
+      for (Map.Entry<String, List<String>> extraHeader : extraHeaders.entrySet()) {
+        newExtraHeadersBuilder.putAll(extraHeader.getKey(), extraHeader.getValue());
+      }
+    }
+    newExtraHeadersBuilder.putAll(this.extraHeaders);
```

Comment:
This will merge in existing headers, and will do it after new headers, it looks like i is possible only to add more and more headers in callContext. How to remove/replace existing headers? If this.extraHeaders were put before new onews it would allow overwriting old headers with new headers (but only if `Map<String, List<Sring>>` was used as internal representation, since it looks like `ImmutableListMultimap` will just keep adding stuff no matter what).

Reply:
That being said, changing to `Map<String, List<String>>` seems to make it possible to replace headers (but not remove) if we still want to merge `extraHeaders` into `this.extraHeaders` (i.e., copying entries over. But if we do something like `this.extraHeaders = extraHeaders` instead we will not be able to merge `extraHeaders` into existing `this.extraHeaders`. Which way do we want to go?
Also cc/ @pongad

Reply:
A few points here.

1. I think we should put old headers before new ones. I'm not sure if orders matter, but I think it makes sense to be consistent.

2. If you change your internal representation to `ImmutableMap`, you still can't do `ImmutableMap.builder().putAll(this.extraHeaders).putAll(extraHeaders).build()`. `ImmutableMap` only lets you put a key once, and doing it more than once will fail.

You could do this though:
```java
for (Map.Entry<String, String> oldHeader : this.extraHeaders.entries()) {
  if (!extraHeaders.contains(oldHeader.key()) newExtreaheadersBuilder.put(oldHeader);
}
newExtreaheadersBuilder.putAll(extraHeaders);
```

3. I find the idea of overwriting existing headers a little funny as a general case. If it were completely up to me, I'll probably have two methods `withExtraHeader` for appending and `withReplacedHeader` for replacing. I'll let @vam-google and @garrettjonesgoogle decide on this though; you've worked with gax for a lot longer than I have.

4. A bit of a gimmick, but if we want to preserve the current behavior, you can also
`if (extraHeaders == null || extraHeaders.isEmpty()) return this`.

Reply:
I took another look at the specification for `Metadata` (https://grpc.io/grpc-java/javadoc/io/grpc/Metadata.html#put-io.grpc.Metadata.Key-T-) and made sure that `Metadata` allows both duplicate keys and duplicate values for the same key. `Metadata.put(key, value)` preserves this feature and adds `value` to the end if `key` or both `key` and `value` already exist. I'm in favor of making `withExtraHeaders(Map<String, List<String>>)` behave the same as `Metadata.put(key, value)` in this case for consistency, i.e, something like
```
// assuming this.extraHeaders is a ImmutableListMultimap
newExtraHeaders = ImmutableListMultimap.builder()
    .putAll(this.extraHeaders).putAll(extraHeaders).build();

// assuming this.extraHeaders is a ImmutableMap<String, List<String>>
newExtraHeadersBuilder = ImmutableMap.builder();
for (Map.Entry<String, List<String> oldHeader : this.extraHeaders.entrySet()) {
  String key = oldHeader.getKey();
  if (extraHeaders.contains(key) {    
    newExtraHeadersBuilder.put(key, ImmutableList.<String>builder()
                       .addAll(oldHeader.getValue()).addAll(extraHeaders.get(key)).build());
  }
  else {
    newExtraHeadersBuilder.put(key, oldHeader.getValue());
  }
}
for (Map.Entry<String, List<String> header : extraHeaders.entrySet()) {
  if (!this.extraHeaders.contains(header.getKey()) {
    newExtraHeadersBuilder.put(header.getKey(), header.getValue()); 
  } 
}
newExtraHeaders = newExtraHeadersBuilder.build();

```
For the functionality of removing/replacing, `Metadata` itself does not provide methods for replacing but does provide `remove()` and `removeAll()`. I am OK with @pongad 's `withReplacedHeader(Map<> extraHeaders)` (this will presumably replace everything in `this.extraHeaders` with `extraHeaders`). IMO I guess we could also do `withEmptyExtraHeaders()`, which simply removes the current `extraHeaders` entirely (just like `removeAll()`). And I guess there is no need to provide methods for replace a single key-value pair since `Metadata` does not do that either. Do you guys have any thoughts on that? @garrettjonesgoogle @vam-google Thanks!

Reply:
I like the idea of making it consistent with `Metadata` class.